coverage: Emit the filenames section before encoding per-function mappings

Zalathar opened the pull request:
When embedding coverage information in LLVM IR (and ultimately in the resulting binary), there are two main things that each CGU needs to emit:

- A single `__llvm_covmap` record containing a coverage header, which mostly consists of a list of filenames used by the CGU's coverage mappings.
- Several `__llvm_covfun` records, one for each instrumented function, each of which contains the hash of the list of filenames in the header.

There is a kind of loose cyclic dependency between the two: we need the hash of the file table before we can emit the covfun records, but we need to traverse all of the instrumented functions in order to build the file table.

The existing code works by processing the individual functions first. It lazily adds filenames to the file table, and stores the mostly-complete function records in a temporary list. After this it hashes the file table, emits the header (containing the file table), and then uses the hash to emit all of the function records.

This PR reverses that order: first we traverse all of the functions (without trying to prepare their function records) to build a *complete* file table, and then emit it immediately. At this point we have the file table hash, so we can then proceed to build and emit all of the function records, without needing to store them in an intermediate list.

---

Along the way, this PR makes some necessary changes that are also worthwhile in their own right:
- We split `FunctionCoverage` into distinct collector/finished phases, which neatly avoids some borrow-checker hassles when extracting a function's final expression/mapping data.
- We avoid having to re-sort a function's mappings when preparing the list of filenames that it uses.

Zalathar committed:
coverage: Encapsulate local-to-global file mappings

Zalathar committed:
coverage: Encode function mappings without re-sorting them

The main change here is that `VirtualFileMapping` now uses an internal hashmap
to de-duplicate incoming global file IDs. That removes the need for
`encode_mappings_for_function` to re-sort its mappings by filename in order to
de-duplicate them.

(We still de-duplicate runs of identical filenames to save work, but this is
not load-bearing for correctness, so a sort is not necessary.)

Zalathar committed:
coverage: Build the global file table ahead of time

Zalathar committed:
coverage: Split `FunctionCoverage` into distinct collector/finished phases

This gives us a clearly-defined place to run code after the instance's MIR has
been traversed by codegen, but before we emit its `__llvm_covfun` record.

Zalathar committed:
coverage: Fetch expressions and mappings separately

The combined `get_expressions_and_counter_regions` method was an artifact of
having to prepare the expressions and mappings at the same time, to avoid
ownership/lifetime problems with temporary data used by both.

Now that we have an explicit transition from `FunctionCoverageCollector` to the
final `FunctionCoverage`, we can prepare any shared data during that step and
store it in the final struct.

Zalathar committed:
coverage: Emit the filenames section before encoding per-function mappings

Most coverage metadata is encoded into two sections in the final executable.
The `__llvm_covmap` section mostly just contains a list of filenames, while the
`__llvm_covfun` section contains encoded coverage maps for each instrumented
function.

The catch is that each per-function record also needs to contain a hash of the
filenames list that it refers to. Historically this was handled by assembling
most of the per-function data into a temporary list, then assembling the
filenames buffer, then using the filenames hash to emit the per-function data,
and then finally emitting the filenames table itself.

However, now that we build the filenames table up-front (via a separate
traversal of the per-function data), we can hash and emit that part first, and
then emit each of the per-function records immediately after building. This
removes the awkwardness of having to temporarily store nearly-complete
per-function records.

cjgillot commented on compiler/rustc_codegen_llvm/src/coverageinfo/mapgen.rs:
```diff
@@ -139,40 +146,64 @@ pub fn finalize(cx: &CodegenCx<'_, '_>) {
+        let all_file_names = all_file_names.into_iter().dedup().collect::<Vec<_>>();
+        for file_name in &all_file_names {
+            raw_file_table.entry(*file_name).or_insert_with(|| file_name.as_str());
+        }
```
Can we work directly with the `FxIndexSet<Symbol>`? Or is it too slow to access the interned strings?

Zalathar replied:
Yeah, that's probably fine.

cjgillot commented on compiler/rustc_codegen_llvm/src/coverageinfo/map_data.rs:
```diff
@@ -165,72 +162,84 @@ impl<'tcx> FunctionCoverage<'tcx> {
+
+impl<'tcx> FunctionCoverage<'tcx> {
+    fn from_collector(collector: FunctionCoverageCollector<'tcx>) -> Self {
+        let zero_expressions = collector.identify_zero_expressions();
+        let FunctionCoverageCollector { function_coverage_info, is_used, counters_seen, .. } =
+            collector;
+
+        Self { function_coverage_info, is_used, counters_seen, zero_expressions }
```
Nit: should this go in `into_finished` and remove this method?

Zalathar commented:
@rustbot label +A-code-coverage

Zalathar commented:
Addressed review feedback [(diff)](https://github.com/rust-lang/rust/compare/ba3859b77452b896c4606533669a10ff67abe068..c46abe4319e57b79561209d4051c2cd04f5fecd2).

cjgillot commented:
@bors r+

bors commented:
:pushpin: Commit c46abe4319e57b79561209d4051c2cd04f5fecd2 has been approved by `cjgillot`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cjgillot c46abe4319e57b79561209d4051c2cd04f5fecd2 -->
<!-- homu: {"type":"Approved","sha":"c46abe4319e57b79561209d4051c2cd04f5fecd2","approver":"cjgillot","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit c46abe4319e57b79561209d4051c2cd04f5fecd2 with merge 0611ae28d71fef35915319642d447546fe712ad3...
<!-- homu: {"type":"BuildStarted","head_sha":"c46abe4319e57b79561209d4051c2cd04f5fecd2","merge_sha":"0611ae28d71fef35915319642d447546fe712ad3"} -->

rust-log-analyzer commented:
The job **`aarch64-gnu`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/6603396636/job/17936457344) [(plain)](https://github.com/rust-lang-ci/rust/commit/0611ae28d71fef35915319642d447546fe712ad3/checks/17936457344/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
---- [run-make] tests/run-make/coverage-llvmir stdout ----

error: make failed
status: exit status: 2
command: cd "/checkout/tests/run-make/coverage-llvmir" && env -u CARGO_MAKEFLAGS -u MAKEFLAGS -u MFLAGS -u RUSTFLAGS AR="ar" CC="cc -ffunction-sections -fdata-sections -fPIC" CXX="c++ -ffunction-sections -fdata-sections -fPIC" HOST_RPATH_DIR="/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/lib" LD_LIB_PATH_ENVVAR="LD_LIBRARY_PATH" LLVM_BIN_DIR="/checkout/obj/build/aarch64-unknown-linux-gnu/llvm/build/bin" LLVM_COMPONENTS="aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen codegentypes core coroutines coverage csky cskyasmparser cskycodegen cskydesc cskydisassembler cskyinfo debuginfobtf debuginfocodeview debuginfodwarf debuginfogsym debuginfologicalview debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker dwarflinkerparallel dwp engine executionengine extensions filecheck frontendhlsl frontendopenacc frontendopenmp fuzzercli fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interfacestub interpreter ipo irprinter irreader jitlink libdriver lineeditor linker loongarch loongarchasmparser loongarchcodegen loongarchdesc loongarchdisassembler loongarchinfo lto m68k m68kasmparser m68kcodegen m68kdesc m68kdisassembler m68kinfo mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts objcopy object objectyaml option orcjit orcshared orctargetprocess passes powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvtargetmca runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target targetparser textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo webassemblyutils windowsdriver windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86targetmca xray" LLVM_FILECHECK="/checkout/obj/build/aarch64-unknown-linux-gnu/llvm/build/bin/FileCheck" PYTHON="/usr/bin/python3" RUSTC="/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" RUSTDOC="/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustdoc" RUST_BUILD_STAGE="stage2-aarch64-unknown-linux-gnu" RUST_DEMANGLER="/checkout/obj/build/aarch64-unknown-linux-gnu/stage2-tools-bin/rust-demangler" S="/checkout" TARGET="aarch64-unknown-linux-gnu" TARGET_RPATH_DIR="/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/lib/rustlib/aarch64-unknown-linux-gnu/lib" TMPDIR="/checkout/obj/build/aarch64-unknown-linux-gnu/test/run-make/coverage-llvmir/coverage-llvmir" "make"
# Compile the test program with non-experimental coverage instrumentation, and generate LLVM IR
# Compile the test program with non-experimental coverage instrumentation, and generate LLVM IR
LD_LIBRARY_PATH="/checkout/obj/build/aarch64-unknown-linux-gnu/test/run-make/coverage-llvmir/coverage-llvmir:/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/aarch64-unknown-linux-gnu/stage0-bootstrap-tools/aarch64-unknown-linux-gnu/release/deps:/checkout/obj/build/aarch64-unknown-linux-gnu/stage0/lib" '/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc' --out-dir /checkout/obj/build/aarch64-unknown-linux-gnu/test/run-make/coverage-llvmir/coverage-llvmir -L /checkout/obj/build/aarch64-unknown-linux-gnu/test/run-make/coverage-llvmir/coverage-llvmir  -Ainternal_features ../coverage-llvmir/testprog.rs \
  -Cinstrument-coverage \
  --emit=llvm-ir
cat "/checkout/obj/build/aarch64-unknown-linux-gnu/test/run-make/coverage-llvmir/coverage-llvmir"/testprog.ll | \
  "/checkout/obj/build/aarch64-unknown-linux-gnu/llvm/build/bin/FileCheck" ../coverage-llvmir/filecheck.testprog.txt -check-prefixes=CHECK -DDEFINE_INTERNAL='define internal' -DCOMDAT_IF_SUPPORTED=', comdat' -DINSTR_PROF_DATA='__llvm_prf_data' -DINSTR_PROF_NAME='__llvm_prf_names' -DINSTR_PROF_CNTS='__llvm_prf_cnts' -DINSTR_PROF_VALS='__llvm_prf_vals' -DINSTR_PROF_VNODES='__llvm_prf_vnds' -DINSTR_PROF_COVMAP='__llvm_covmap' -DINSTR_PROF_COVFUN='__llvm_covfun' -DINSTR_PROF_ORDERFILE='__llvm_orderfile'
--- stderr -------------------------------
--- stderr -------------------------------
../coverage-llvmir/filecheck.testprog.txt:9:8: error: CHECK: expected string not found in input
CHECK: @__llvm_coverage_mapping = private constant
<stdin>:73:266: note: scanning from here
<stdin>:73:266: note: scanning from here
@__covrec_A2CB80454FA0AB25u = linkonce_odr hidden constant <{ i64, i32, i64, i64, [9 x i8] }> <{ i64 -6716133384115934427, i32 9, i64 -7006449540818706878, i64 5760647764442993253, [9 x i8] c"\01\01\00\01\01\0C\01\05\02" }>, section "__llvm_covfun", comdat, align 8
<stdin>:74:293: note: possible intended match here
<stdin>:74:293: note: possible intended match here
@__covrec_138797266E6ABDC1u = linkonce_odr hidden constant <{ i64, i32, i64, i64, [28 x i8] }> <{ i64 1407259599893609921, i32 28, i64 -6557364581736350823, i64 5760647764442993253, [28 x i8] c"\01\01\02\01\05\05\02\04\01\13\01\04\13\05\05\09\00\18\02\01\06\00\07\07\01\01\00\02" }>, section "__llvm_covfun", comdat, align 8

Input file: <stdin>
Check file: ../coverage-llvmir/filecheck.testprog.txt


-dump-input=help explains the following input dump.
Input was:
<<<<<<
           .
           .
           .
           .
          68: @alloc_b37d567b61b40850b7d2688454dbb1e3 = private unnamed_addr constant <{ ptr, [8 x i8], ptr, [8 x i8], ptr, [8 x i8], ptr, [8 x i8] }> <{ ptr @alloc_513570631223a12912d85da2bec3b15a, [8 x i8] zeroinitializer, ptr @alloc_513570631223a12912d85da2bec3b15a, [8 x i8] zeroinitializer, ptr @alloc_513570631223a12912d85da2bec3b15a, [8 x i8] zeroinitializer, ptr @alloc_49a1e817e911805af64bbc7efb390101, [8 x i8] c"\01\00\00\00\00\00\00\00" }>, align 8 
          69: @alloc_0fd49fb9c87d0ea55f07188f9eff834d = private unnamed_addr constant <{ [4 x i8] }> <{ [4 x i8] c" ***" }>, align 1 
          70: @alloc_9db824fd26015da05d288b42181536c6 = private unnamed_addr constant <{ [4 x i8] }> <{ [4 x i8] c"*** " }>, align 1 
          71: @alloc_3edf33e52be22c945259abc2c0759d11 = private unnamed_addr constant <{ [19 x i8] }> <{ [19 x i8] c"wrapped result is: " }>, align 1 
          72: @__llvm_coverage_mapping = private constant { { i32, i32, i32, i32 }, [75 x i8] } { { i32, i32, i32, i32 } { i32 0, i32 75, i32 0, i32 5 }, [75 x i8] c"\02H\00(/checkout/tests/run-make/coverage-llvmir\1E../coverage-llvmir/testprog.rs" }, section "__llvm_covmap", align 8 
          73: @__covrec_A2CB80454FA0AB25u = linkonce_odr hidden constant <{ i64, i32, i64, i64, [9 x i8] }> <{ i64 -6716133384115934427, i32 9, i64 -7006449540818706878, i64 5760647764442993253, [9 x i8] c"\01\01\00\01\01\0C\01\05\02" }>, section "__llvm_covfun", comdat, align 8 
check:9'0                                                                                                                                                                                                                                                                              X error: no match found
          74: @__covrec_138797266E6ABDC1u = linkonce_odr hidden constant <{ i64, i32, i64, i64, [28 x i8] }> <{ i64 1407259599893609921, i32 28, i64 -6557364581736350823, i64 5760647764442993253, [28 x i8] c"\01\01\02\01\05\05\02\04\01\13\01\04\13\05\05\09\00\18\02\01\06\00\07\07\01\01\00\02" }>, section "__llvm_covfun", comdat, align 8 
check:9'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:9'1                                                                                                                                                                                                                                                                                                         ?                                 possible intended match
          75: @__covrec_AAF7C95823D55CB4u = linkonce_odr hidden constant <{ i64, i32, i64, i64, [9 x i8] }> <{ i64 -6127207387618779980, i32 9, i64 -3200427416441104464, i64 5760647764442993253, [9 x i8] c"\01\01\00\01\01$7\00^" }>, section "__llvm_covfun", comdat, align 8 
check:9'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          76: @__covrec_1D9B895249AD0C75u = linkonce_odr hidden constant <{ i64, i32, i64, i64, [9 x i8] }> <{ i64 2133449835006463093, i32 9, i64 -7749979286589876189, i64 5760647764442993253, [9 x i8] c"\01\01\00\01\01\01\01\04\02" }>, section "__llvm_covfun", comdat, align 8 
check:9'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          77: @__covrec_36725C27D3039CB5u = linkonce_odr hidden constant <{ i64, i32, i64, i64, [9 x i8] }> <{ i64 3923299551506832565, i32 9, i64 -427819517993304365, i64 5760647764442993253, [9 x i8] c"\01\01\00\01\01\07\01\03\02" }>, section "__llvm_covfun", comdat, align 8 
check:9'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          78: @__covrec_53F9FD11B26D788Au = linkonce_odr hidden constant <{ i64, i32, i64, i64, [43 x i8] }> <{ i64 6051145826798762122, i32 43, i64 -3200427416441104464, i64 5760647764442993253, [43 x i8] c"\01\01\02\01\05\05\02\07\01\1C\01\04\13\05\04\14\012\05\01V\012\05\01V\01\06\02\01\0C\01/\02\01^\01\06\07\02\01\00\02" }>, section "__llvm_covfun", comdat, align 8 
check:9'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          79: @__covrec_1CE7CA97E6F1AF41u = linkonce_odr hidden constant <{ i64, i32, i64, i64, [9 x i8] }> <{ i64 2082856106445614913, i32 9, i64 -7006449540818706878, i64 5760647764442993253, [9 x i8] c"\01\01\00\01\01\0C\01\05\02" }>, section "__llvm_covfun", comdat, align 8 
check:9'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           .
           .
>>>>>>
>>>>>>
make: *** [Makefile:60: test_llvm_ir] Error 1
Build completed unsuccessfully in 0:48:21



```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6603396636/job/17936884760)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/6603396636/job/17936884760","builder_name":"checks-actions"} -->

Zalathar commented:
Fixed outdated test [(diff)](https://github.com/rust-lang/rust/compare/c46abe4319e57b79561209d4051c2cd04f5fecd2..6af9fef08590fd499370a2f6cbbae9ceacf15336).

@rustbot ready

cjgillot commented:
@bors r+

bors commented:
:pushpin: Commit 6af9fef08590fd499370a2f6cbbae9ceacf15336 has been approved by `cjgillot`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cjgillot 6af9fef08590fd499370a2f6cbbae9ceacf15336 -->
<!-- homu: {"type":"Approved","sha":"6af9fef08590fd499370a2f6cbbae9ceacf15336","approver":"cjgillot","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
